collect_app/src/test/java/org/odk/collect/android/utilities/CopyImageAndApplyExifRotationTest.kt

```diff
@@ -0,0 +1,150 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.odk.collect.android.utilities
+
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import androidx.exifinterface.media.ExifInterface
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import timber.log.Timber
+import java.io.File
+import java.io.IOException
+
+@RunWith(AndroidJUnit4::class)
+class CopyImageAndApplyExifRotationTest {
+
+    private lateinit var sourceFile: File
+    private lateinit var destinationFile: File
+    private lateinit var attributes: MutableMap<String, String>
+
+    @Before
+    fun setup() {
+        sourceFile = createTempImageFile("source")
+        destinationFile = createTempImageFile("destination")
+        attributes = HashMap()
+    }
+
+    @Test
+    fun copyAndRotateImageNinety() {
+        attributes[ExifInterface.TAG_ORIENTATION] = ExifInterface.ORIENTATION_ROTATE_90.toString()
+        saveTestBitmapToFile(sourceFile.absolutePath, 3000, 4000, attributes)
+        FileUtils.copyImageAndApplyExifRotation(sourceFile, destinationFile)
+        val image = FileUtils.getBitmap(destinationFile.absolutePath, BitmapFactory.Options())
+
+        assertEquals(4000, image.width)
+        assertEquals(3000, image.height)
+        verifyNoExifOrientationInDestinationFile(destinationFile.absolutePath)
+    }
+
+    @Test
+    fun copyAndRotateImageTwoSeventy() {
+        attributes[ExifInterface.TAG_ORIENTATION] = ExifInterface.ORIENTATION_ROTATE_270.toString()
+        saveTestBitmapToFile(sourceFile.absolutePath, 3000, 4000, attributes)
+        FileUtils.copyImageAndApplyExifRotation(sourceFile, destinationFile)
+        val image = FileUtils.getBitmap(destinationFile.absolutePath, BitmapFactory.Options())
+
+        assertEquals(4000, image.width)
+        assertEquals(3000, image.height)
+        verifyNoExifOrientationInDestinationFile(destinationFile.absolutePath)
+    }
+
+    @Test
+    fun copyAndRotateImageOneEighty() {
+        attributes[ExifInterface.TAG_ORIENTATION] = ExifInterface.ORIENTATION_ROTATE_180.toString()
+        saveTestBitmapToFile(sourceFile.absolutePath, 3000, 4000, attributes)
+        FileUtils.copyImageAndApplyExifRotation(sourceFile, destinationFile)
+        val image = FileUtils.getBitmap(destinationFile.absolutePath, BitmapFactory.Options())
+
+        assertEquals(3000, image.width)
```

Comment:
This test still passes if I change the `ORIENTATION_ROTATE_180` branch in the implementation to perform a rotation of `0` - 180 is annoying as the image dimensions end up the same. We technically have a similar problem for 90 and 270 as they could be flipped and the tests would still pass. 

Are we able to verify these by creating a Bitmap with a single filled pixel and checking where it ends up? I can imagine that might not be worth it wanted to discuss. Maybe there is a better way that you can think of.

Reply:
I agree and I improved that test. It required moving it to `androidTest` because when you create/open bitmaps in unit test it returns empty bitmaps (the size is correct but its empty/transparent).